libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/QnAMaker/QnAMakerRecognizer.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Licensed under the MIT License.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Linq;
+using System.Net.Http;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.AI.QnA;
+using Microsoft.Bot.Expressions;
+using Microsoft.Bot.Schema;
+using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive.QnA.Recognizers
+{
+    /// <summary>
+    /// IRecognizer implementation which uses QnAMaker KB to identify intents.
+    /// </summary>
+    public class QnAMakerRecognizer : InputRecognizer
+    {
+        [JsonProperty("$kind")]
+        public const string DeclarativeType = "Microsoft.QnAMakerRecognizer";
+
+        public const string QnAMatchIntent = "QnAMatch";
+
+        private Expression knowledgebaseIdExpression;
+        private Expression endpointkeyExpression;
+        private Expression hostnameExpression;
+
+        public QnAMakerRecognizer()
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the KnowledgeBase Id of your QnA Maker KnowledgeBase.
+        /// </summary>
+        /// <value>
+        /// The knowledgebase Id.
+        /// </value>
+        [JsonProperty("knowledgeBaseId")]
+        public string KnowledgeBaseId
+        {
+            get { return knowledgebaseIdExpression?.ToString(); }
+            set { knowledgebaseIdExpression = value != null ? new ExpressionEngine().Parse(value) : null; }
```

Comment:
>ExpressionEngine [](start = 66, length = 16)

This is a little brittle in that if people extend for their own functions this will not work. If you look at the selector stuff in the form PR you can see how I handled that.

Reply:
Filed seperate issue to do a sweep and catch all ofthese. [https://github.com/microsoft/botbuilder-dotnet/issues/3185](https://github.com/microsoft/botbuilder-dotnet/issues/3185)

---
In reply to: [360620251](https://github.com/microsoft/botbuilder-dotnet/pull/3174#discussion_r360620251) [](ancestors = 360620251)